### PR TITLE
Add null qubit

### PR DESCRIPTION
**Pull request by AmintorDusko**
**Context:** Implement a Catalyst runtime plugin that mocks out all functions in the QuantumDevice interface and implement a PennyLane device plugin (implemented in [qjit_device.py](https://github.com/PennyLaneAI/catalyst/blob/main/frontend/catalyst/device/qjit_device.py)) providing access to the Null runtime device.

**Description of the Change:**
- [x] Rename the device from "dummy" to "null".
- [x] Make sure the device is up to date with the latest version of the QuantumDevice interface.
- [x] Ensure proper unit testing of the device in C++.
- [x] Package and ship this plugin for both development builds and wheels.

**Benefits:**
Such a device will be useful for a few reasons:

* "execute" programs without needing to install additional plugins
* Simplify and speed up integration tests when numerical results aren't needed
* Benchmark programs while excluding the cost of simulation -> exposes the compilation and infrastructure overhead
* Can also act as an instruction tracer / program printer

**Possible Drawbacks:** It is a surprise.

**Related GitHub Issues:**

[sc-72829] [sc-72830]

### CODE REVIEW

**Review by erick-xanadu — COMMENTED** [review]
Hi @AmintorDusko , everything I've seen looks good. Whenever you feel ready to open it for review, just change it so that it is not in draft mode. Cheers!

File: frontend/test/lit/test_decomposition.py [hunk 1]
```diff
@@ -70,9 +70,11 @@ def get_c_interface():
             """
             system_extension = ".dylib" if platform.system() == "Darwin" else ".so"
             lib_path = (
-                get_lib_path("runtime", "RUNTIME_LIB_DIR") + "/librtd_dummy" + system_extension
+                get_lib_path("runtime", "RUNTIME_LIB_DIR")
+                + "/librtd_null_device"
+                + system_extension
             )
-            return "dummy.remote", lib_path
+            return "NullDevice", lib_path
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply frontend/test/lit/test_decomposition.py:77]: It can be 'null.device'—there is nothing against it. I was just trying to normalize the names. In some places, we had CustomDevice, in other places `dummy.remote`. In the Lightning repo we have 'LightningKokkosSimulator' for LKokkos. I don't know the relevance of the naming here for Catalyst, so if you say it is better to make it `null.device` I can do that. Just let me know. :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (erick-xanadu) [reply frontend/test/lit/test_decomposition.py:77]: No, I just haven't seen this part of the code in a long time, so if you say it is more common to have another string, then let's keep your original name. Cheers! :)

File: frontend/test/lit/test_device_api.py [hunk 2]
```diff
@@ -65,8 +65,9 @@ def preprocess(self, execution_config: Optional[ExecutionConfig] = None):
 def test_circuit():
     """Test a circuit compilation to MLIR when using the new device API."""
 
-    # CHECK:    quantum.device["[[PATH:.*]]librtd_lightning.{{so|dylib}}", "dummy.remote", "{'shots': 2048}"]
-    dev = DummyDevice(wires=2, shots=2048)
+    # CHECK
+    # quantum.device["[[PATH:.*]]librtd_null_device.{{so|dylib}}", "NullDevice", "{'shots': 2048}"]
```
Comment (erick-xanadu) [inline frontend/test/lit/test_device_api.py:69]: Please revert these changes. These CHECK: comments are the actual test.

```suggestion
    # CHECK:    quantum.device["[[PATH:.*]]librtd_lightning.{{so|dylib}}", "dummy.remote", "{'shots': 2048}"]
```

File: frontend/test/lit/test_device_api.py [hunk 3]
```diff
@@ -90,8 +91,9 @@ def test_preprocess():
     using the new device API.
     TODO: we need to readd the two check-not once we accept the device preprocessing."""
 
-    # CHECK:    quantum.device["[[PATH:.*]]librtd_lightning.{{so|dylib}}", "dummy.remote", "{'shots': 2048}"]
-    dev = DummyDevice(wires=2, shots=2048)
+    # CHECK
+    # quantum.device["[[PATH:.*]]librtd_null_device.{{so|dylib}}", "NullDevice", "{'shots': 2048}"]
```
Comment (erick-xanadu) [inline frontend/test/lit/test_device_api.py:95]: Same here: it should be one line.

```suggestion
    # CHECK:    quantum.device["[[PATH:.*]]librtd_null_device.{{so|dylib}}", "NullDevice", "{'shots': 2048}"]
```

File: frontend/test/lit/test_device_api.py (hunk 2 above)
Comment (AmintorDusko) [reply frontend/test/lit/test_device_api.py:69]: Oops. Sorry. But the codefactor is complaining about the line size.

File: frontend/test/lit/test_decomposition.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/test/lit/test_decomposition.py:77]: Yes I think `*.*` is the typical syntax for the device short name that users will use to construct it in Python. But here the provided name is used as a C++ class or function (the library entry point), so the dot might not even be legal?

File: frontend/test/lit/test_device_api.py [hunk 4]
```diff
@@ -28,10 +28,10 @@
 from catalyst.compiler import get_lib_path
 
 
-class DummyDevice(Device):
-    """A dummy device from the device API."""
+class NullDevice(Device):
```
Comment (dime10) [inline frontend/test/lit/test_device_api.py:31]: For these tests we do might want to create a custom class because we are testing the device api, in which case we might want to indicate this in the name though, `CustomDevice` instead of `NullDevice` for example.

**Review by dime10 — COMMENTED** [review]
Thanks for the PR @AmintorDusko, left an early review regarding the direction we want to go with the device but looking good so far 👍

File: Makefile [hunk 5]
```diff
@@ -112,8 +112,8 @@ dialects:
 runtime:
 	$(MAKE) -C runtime runtime
 
-dummy_device:
-	$(MAKE) -C runtime dummy_device
+null_device:
+	$(MAKE) -C runtime null_device
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply Makefile:116]: Sure!

File: frontend/catalyst/_configuration.py [hunk 6]
```diff
@@ -1,6 +1 @@
-""" When catalyst is packaged into a python wheel
-    the contents of this file will be overridden with:
-    INSTALLED = True
-"""
-
-INSTALLED = False
+INSTALLED = True
```
Comment (AmintorDusko) [inline frontend/catalyst/_configuration.py:1]:
```suggestion
""" When catalyst is packaged into a python wheel
    the contents of this file will be overridden with:
    INSTALLED = True
"""

INSTALLED = False
```

File: frontend/test/pytest/test_custom_devices.py [hunk 7]
```diff
@@ -209,26 +209,26 @@ def get_c_interface():
...
 
-    class DummyDevice(qml.devices.QubitDevice):
-        """Dummy Device"""
+    class CustomDevice(qml.devices.QubitDevice):
+        """Custom Device"""
 
-        name = "Dummy Device"
-        short_name = "dummy.device"
+        name = "Custom Qubit"
+        short_name = "null.qubit"
```
Comment (erick-xanadu) [inline frontend/test/pytest/test_custom_devices.py:231]:
```suggestion
        short_name = "custom.qubit"
```

File: frontend/test/pytest/test_custom_devices.py [hunk 8]
```diff
@@ -184,11 +184,11 @@ def f():
...
```
Comment (erick-xanadu) [inline frontend/test/pytest/test_custom_devices.py:191]:
```suggestion
        short_name = "custom.qubit"
```

**Review by erick-xanadu — CHANGES_REQUESTED** [review]
The tests you are failing is because the short_name is being matched in the list of `SUPPORTED_RT_DEVICES`

```
SUPPORTED_RT_DEVICES = {
    "null.qubit": ("NullQubit", "librtd_null_qubit"),
```

I think if you just change the name, the tests will pass again.

File: runtime/lib/capi/ExecutionContext.hpp [hunk 9]
```diff
@@ -182,7 +184,11 @@ class RTDevice {
         // Besides, this provides support for runtime device (RTD) libraries added to the system
         // path. This maintains backward compatibility for specifying a device using its name.
         // TODO: This support may need to be removed after updating the C++ unit tests.
-        if (rtd_lib == "lightning.qubit" || rtd_lib == "lightning.kokkos") {
+        if (rtd_lib == "null.qubit") {
+            rtd_name = "NullQubit";
+            _complete_dylib_os_extension(rtd_lib, "null_qubit");
+        }
+        else if (rtd_lib == "lightning.qubit" || rtd_lib == "lightning.kokkos") {
```
Comment (dime10) [inline runtime/lib/capi/ExecutionContext.hpp:191]: Speaking off, lightning kokkos no longer lives here so it can be removed from the list right?

File: runtime/lib/capi/ExecutionContext.hpp [hunk 10]
```diff
@@ -182,7 +184,11 @@ class RTDevice {
         // Besides, this provides support for runtime device (RTD) libraries added to the system
         // path. This maintains backward compatibility for specifying a device using its name.
         // TODO: This support may need to be removed after updating the C++ unit tests.
```
Comment (dime10) [inline runtime/lib/capi/ExecutionContext.hpp:186]: Do you see a way of doing away with these special cases somehow? It seems they only exist for testing.

File: runtime/tests/Test_NullQubit.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
+
+// Copyright 2023 Xanadu Quantum Technologies Inc.
```
Comment (dime10) [inline runtime/tests/Test_NullQubit.cpp:2]:
```suggestion
// Copyright 2024 Xanadu Quantum Technologies Inc.
```

File: runtime/tests/TestUtils.hpp [hunk 12]
```diff
@@ -24,18 +24,35 @@
...
 #include "LightningKokkosSimulator.hpp"
 using SimTypes = std::tuple<Catalyst::Runtime::Simulator::LightningSimulator,
                             Catalyst::Runtime::Simulator::LightningKokkosSimulator>;
 #else
 using SimTypes = std::tuple<Catalyst::Runtime::Simulator::LightningSimulator>;
 #endif
+#else
+#if __has_include("LightningKokkosSimulator.hpp")
+#include "LightningKokkosSimulator.hpp"
+using SimTypes = std::tuple<Catalyst::Runtime::Simulator::LightningKokkosSimulator>;
```
Comment (dime10) [inline runtime/tests/TestUtils.hpp:53]: Is this still relevant with kokkos gone?

File: runtime/tests/Test_NullQubit.cpp
Comment (dime10) [inline runtime/tests/Test_NullQubit.cpp:1]: You mentioned there were existing tests for the dummy device, where are those?

**Review by dime10 — COMMENTED** [review]
Thanks @AmintorDusko for contributing this PR. I don't think anything needs a direction change, the biggest challenge is that there seems to be a lot of clean up work required (sorry about that!).

In the description you ticked off:
> Make sure the device is up to date with the latest version of the QuantumDevice interface.

But I don't see any changes to the device implementation. I think we've added at least:
- SetDevicePRNG
- SetState/SetBasisState

We might also want to give the toml file a quick lookover. I think at the time it was meant to copy the lightning.qubit device. I'm not sure if that is still what we want to do or follow a different strategy 🤔 @erick-xanadu do you have any thoughts here? 
At the very least we'll need to add the StatePrep support flag to it.

**Comment by erick-xanadu** [comment]
>  🤔 @erick-xanadu do you have any thoughts here?

I think it should follow lightning's toml file as closely as possible. We don't want to see behaviour X in lightning and then try to debug it with null qubit and then see behaviour Y. (E.g., gates are decomposed differently).

File: runtime/tests/TestUtils.hpp (hunk 12 above)
Comment (AmintorDusko) [reply runtime/tests/TestUtils.hpp:53]: removed!

File: runtime/tests/Test_NullQubit.cpp
Comment (AmintorDusko) [reply runtime/tests/Test_NullQubit.cpp:1]: I think these are all. I will look better into this.

File: frontend/test/pytest/test_config_functions.py [hunk 13]
```diff
@@ -35,8 +35,8 @@
 class DeviceToBeTested(qml.devices.QubitDevice):
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply frontend/test/pytest/test_config_functions.py:35]: removed

File: runtime/lib/capi/ExecutionContext.hpp (hunk 9 above)
Comment (AmintorDusko) [reply runtime/lib/capi/ExecutionContext.hpp:191]: removed

File: frontend/test/lit/test_device_api.py [hunk 14]
```diff
@@ -65,8 +65,8 @@ def preprocess(self, execution_config: Optional[ExecutionConfig] = None):
 def test_circuit():
     """Test a circuit compilation to MLIR when using the new device API."""
 
-    # CHECK:    quantum.device["[[PATH:.*]]librtd_lightning.{{so|dylib}}", "dummy.remote", "{'shots': 2048}"]
-    dev = DummyDevice(wires=2, shots=2048)
+    # CHECK:    quantum.device["[[PATH:.*]]librtd_null_qubit.{{so|dylib}}", "NullQubit", "{'shots': 2048}"]
+    dev = NullQubit(wires=2, shots=2048)
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply frontend/test/lit/test_device_api.py]: Could you please expand on that?

File: frontend/test/lit/test_device_api.py [hunk 15]
```diff
@@ -90,8 +90,8 @@ def test_preprocess():
     using the new device API.
     TODO: we need to readd the two check-not once we accept the device preprocessing."""
 
-    # CHECK:    quantum.device["[[PATH:.*]]librtd_lightning.{{so|dylib}}", "dummy.remote", "{'shots': 2048}"]
-    dev = DummyDevice(wires=2, shots=2048)
+    # CHECK:    quantum.device["[[PATH:.*]]librtd_null_qubit.{{so|dylib}}", "NullQubit", "{'shots': 2048}"]
+    dev = NullQubit(wires=2, shots=2048)
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply frontend/test/lit/test_device_api.py]: Done!

File: frontend/test/pytest/test_custom_devices.py [hunk 16]
```diff
@@ -133,11 +133,11 @@
...
+    class NullQubit(qml.devices.QubitDevice):
+        """Null Qubit"""
 
-        name = "Dummy Device"
-        short_name = "dummy.device"
+        name = "Null Qubit"
+        short_name = "null.qubit"
         pennylane_requires = "0.33.0"
         version = "0.0.1"
         author = "Dummy"
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply frontend/test/pytest/test_custom_devices.py:147]: We still need it, but I added the config file and local tests are going smoothly now.

File: frontend/test/pytest/test_measurement_transforms.py [hunk 17]
```diff
@@ -44,10 +44,10 @@
 # pylint: disable=attribute-defined-outside-init
 
 
-class DummyDevice(Device):
-    """A dummy device from the device API."""
+class NullQubit(Device):
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply frontend/test/pytest/test_measurement_transforms.py:47]: Yes, nice catch. I will turn it into a custom device. Thanks.

File: frontend/test/lit/test_device_api.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply frontend/test/lit/test_device_api.py]: Let me know if you want me to update something here.

File: frontend/test/pytest/test_custom_devices.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/test/pytest/test_custom_devices.py:144]: What do we need them for?

File: frontend/test/pytest/test_preprocess.py [hunk 18]
```diff
@@ -147,12 +111,50 @@ def decomposition(self):
...
+        """
+        system_extension = ".dylib" if platform.system() == "Darwin" else ".so"
+        lib_path = (
+            get_lib_path("runtime", "RUNTIME_LIB_DIR") + "/librtd_null_qubit" + system_extension
+        )
+        return "dummy.remote", lib_path
+
+    def execute(self, circuits, execution_config):
+        """Execution."""
+        return circuits, execution_config
```
Comment (dime10) [inline frontend/test/pytest/test_preprocess.py:140]: This implementation doesn't make any sense
```suggestion
    def execute(self, circuits, execution_config):
        """Execution."""
        raise NotImplementedError
```

File: frontend/test/pytest/test_custom_devices.py [hunk 19]
```diff
@@ -22,6 +22,7 @@
 from catalyst.compiler import get_lib_path
 from catalyst.device import extract_backend_info, get_device_capabilities
 from catalyst.utils.exceptions import CompileError
+from frontend.test.conftest import CONFIG_CUSTOM_DEVICE
```
Comment (dime10) [inline frontend/test/pytest/test_custom_devices.py:25]: This import looks a bit strange to me, does that work generally? It seems like it might only work if launching python from the root directory

**Review by dime10 — APPROVED** [review]
Thanks Amintor, I'll give conditional approval here pending resolution of the last comments.

File: runtime/lib/capi/ExecutionContext.hpp (hunk 10 above)
Comment (dime10) [reply runtime/lib/capi/ExecutionContext.hpp:186]: Was just curious since you are probably more familiar with C++ testing than us, but feel free to resolve if you don't have any ideas here.

File: frontend/test/lit/test_device_api.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/test/lit/test_device_api.py]: Looks like you updated it (at least the C++ device identifier field), so this should be good 👍

File: runtime/tests/CMakeLists.txt [hunk 20]
```diff
@@ -24,6 +24,22 @@ list(APPEND CMAKE_MODULE_PATH ${catch2_SOURCE_DIR}/contrib)
 include(CTest)
 include(Catch)
 
+add_executable(runner_tests_qir_runtime runner_main.cpp)
+
+# To avoid link to libpython, we use pybind11::module interface library.
+target_link_libraries(runner_tests_qir_runtime PRIVATE
+    Catch2::Catch2
+    pybind11::embed
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply runtime/tests/CMakeLists.txt:32]: I copied and pasted it from line 96. I didn't have too much time to think here.

File: runtime/lib/capi/ExecutionContext.hpp (hunk 10 above)
Comment (AmintorDusko) [reply runtime/lib/capi/ExecutionContext.hpp:186]: Yes, this is a nice question. I thought about that but I have no idea.

File: frontend/test/pytest/test_custom_devices.py (hunk 19 above)
Comment (AmintorDusko) [reply frontend/test/pytest/test_custom_devices.py]: Let's go with a safer solution. Thanks for pointing this out.

**Review by erick-xanadu — APPROVED** [review]

File: frontend/test/pytest/test_custom_devices.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply frontend/test/pytest/test_custom_devices.py:144]: The new API needs this. Right?

[the remark this answers is not in the document]
Comment (dime10) [reply frontend/test/pytest/test_custom_devices.py:144]: The new API needs `operations` and `observables` fields?
